File: src/RemoteDesk.Android/app/src/main/java/com/remotedesk/agent/RelayPathStability.java
package com.remotedesk.agent;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Same millisecond-based policy as Windows/Python. Never migrates live sockets. */
final class RelayPathStability {
    static final long PROBE_INTERVAL_MS = 30_000, MINIMUM_HOLD_MS = 180_000, HISTORY_MS = 600_000;
    private static final class Sample {
        final long at; final double value;
        Sample(long at, double value) { this.at = at; this.value = value; }
    }
    private final Map<String, ArrayDeque<Sample>> samples = new HashMap<>();
    private String preferred, challenger;
    private long selectedAt, lastUse, lastProbe = Long.MIN_VALUE, lastRound = Long.MIN_VALUE;
    private int wins;

    synchronized String preferred(long now) {
        if (preferred != null && now - lastUse >= HISTORY_MS) {
            preferred = null; resetChallenge(); samples.clear();
        }
        lastUse = now;
        return preferred;
    }

    synchronized void connected(String path, long now, String failedPreferred) {
        lastUse = now;
        if (preferred == null || preferred.equals(failedPreferred)) select(path, now);
    }

    synchronized boolean beginProbe(long now) {
        if (lastProbe != Long.MIN_VALUE && now - lastProbe < PROBE_INTERVAL_MS) return false;
        lastProbe = now;
        return true;
    }

    synchronized void observeRound(long now, Map<String, Double> observations) {
        if (lastRound != Long.MIN_VALUE && now - lastRound < PROBE_INTERVAL_MS) return;
        lastRound = now;
        for (Map.Entry<String, Double> entry : observations.entrySet()) {
            ArrayDeque<Sample> history = samples.computeIfAbsent(entry.getKey(), key -> new ArrayDeque<>());
            while (!history.isEmpty() && (history.size() >= 8 || now - history.peek().at >= HISTORY_MS)) history.remove();
            double value = entry.getValue();
            history.add(new Sample(now, Double.isFinite(value) && value >= 0 ? value : Double.POSITIVE_INFINITY));
        }
        if (preferred == null || !observations.containsKey(preferred)) { resetChallenge(); return; }
        double current = score(preferred, now, false), bestScore = Double.POSITIVE_INFINITY;
        String best = null;
        List<String> paths = new ArrayList<>(observations.keySet());
        Collections.sort(paths);
        for (String path : paths) {
            if (path.equals(preferred)) continue;
            double value = score(path, now, true);
            if (value < bestScore && worthSwitching(current, value)) { best = path; bestScore = value; }
        }
        if (best == null) { resetChallenge(); return; }
        wins = best.equals(challenger) ? wins + 1 : 1;
        challenger = best;
        if (wins >= 3 && now - selectedAt >= MINIMUM_HOLD_MS) select(best, now);
    }

    static boolean worthSwitching(double current, double candidate) {
        return Double.isFinite(current) && Double.isFinite(candidate) && candidate >= 0 &&
            current - candidate >= 8 && candidate <= current * .75;
    }

    private double score(String path, long now, boolean candidate) {
        ArrayDeque<Sample> history = samples.get(path);
        if (history == null) return Double.POSITIVE_INFINITY;
        List<Double> values = new ArrayList<>(), ok = new ArrayList<>();
        for (Sample sample : history) if (now - sample.at < HISTORY_MS) {
            values.add(sample.value);
            if (Double.isFinite(sample.value)) ok.add(sample.value);
        }
        if (values.size() < 3 || candidate && ok.size() < 3) return Double.POSITIVE_INFINITY;
        if (candidate) for (int i = values.size() - 3; i < values.size(); i++)
            if (!Double.isFinite(values.get(i))) return Double.POSITIVE_INFINITY;
        if (ok.isEmpty()) return 2000;
        Collections.sort(ok);
        double median = (ok.get((ok.size() - 1) / 2) + ok.get(ok.size() / 2)) / 2;
        // P90 of eight samples is just the maximum. One old spike/failure
        // must not count as several independent wins in overlapping windows.
        double p80 = ok.get((int)Math.ceil(ok.size() * .8) - 1);
        return median + .75 * (p80 - median) + 500.0 * Math.max(0, values.size() - ok.size() - 1) / values.size();
    }

    private void select(String path, long now) { preferred = path; selectedAt = now; resetChallenge(); }
    private void resetChallenge() { challenger = null; wins = 0; }
}

File: src/RemoteDesk/RelayPathStability.cs
namespace RemoteDesk;

// Shared policy semantics with Android/Python: milliseconds, bounded recent
// TLS-handshake observations, and hysteresis for FUTURE connections only.
internal sealed class RelayPathStability
{
    internal const int ProbeIntervalMilliseconds = 30_000;
    internal const int MinimumHoldMilliseconds = 180_000;
    internal const int HistoryMilliseconds = 600_000;
    internal const string SystemPath = "<system>";
    private readonly object _sync = new();
    private readonly Dictionary<string, Queue<(long At, double Value)>> _samples = new();
    private string? _preferred, _challenger;
    private long _selectedAt, _lastUse, _lastProbe = long.MinValue, _lastRound = long.MinValue;
    private int _wins;

    internal string? Preferred(long now)
    {
        lock (_sync)
        {
            if (_preferred is not null && now - _lastUse >= HistoryMilliseconds)
            {
                _preferred = _challenger = null;
                _wins = 0;
                _samples.Clear();
            }
            _lastUse = now;
            return _preferred;
        }
    }

    internal void Connected(string path, long now, string? failedPreferred = null)
    {
        lock (_sync)
        {
            _lastUse = now;
            // A real dial failure bypasses the hold; cancelled race losers and
            // a single slow handshake do not oscillate the cached preference.
            if (_preferred is null || _preferred == failedPreferred)
                Select(path, now);
        }
    }

    internal bool BeginProbe(long now)
    {
        lock (_sync)
        {
            if (_lastProbe != long.MinValue && now - _lastProbe < ProbeIntervalMilliseconds) return false;
            _lastProbe = now;
            return true;
        }
    }

    internal void ObserveRound(long now, IReadOnlyDictionary<string, double> observations)
    {
        lock (_sync)
        {
            if (_lastRound != long.MinValue && now - _lastRound < ProbeIntervalMilliseconds) return;
            _lastRound = now;
            foreach (var (path, value) in observations)
            {
                if (!_samples.TryGetValue(path, out var history)) _samples[path] = history = new();
                while (history.Count > 0 && (history.Count >= 8 || now - history.Peek().At >= HistoryMilliseconds))
                    history.Dequeue();
                history.Enqueue((now, double.IsFinite(value) && value >= 0 ? value : double.PositiveInfinity));
            }
            if (_preferred is null || !observations.ContainsKey(_preferred)) { ResetChallenge(); return; }
            double current = Score(_preferred, now, candidate: false);
            string? best = null;
            double bestScore = double.PositiveInfinity;
            foreach (string path in observations.Keys.Order(StringComparer.Ordinal))
            {
                if (path == _preferred) continue;
                double score = Score(path, now, candidate: true);
                if (score < bestScore && WorthSwitching(current, score)) { best = path; bestScore = score; }
            }
            if (best is null) { ResetChallenge(); return; }
            _wins = best == _challenger ? _wins + 1 : 1;
            _challenger = best;
            if (_wins >= 3 && now - _selectedAt >= MinimumHoldMilliseconds) Select(best, now);
        }
    }

    internal static bool WorthSwitching(double current, double candidate) =>
        double.IsFinite(current) && double.IsFinite(candidate) && candidate >= 0 &&
        current - candidate >= 8 && candidate <= current * .75;

    private double Score(string path, long now, bool candidate)
    {
        if (!_samples.TryGetValue(path, out var history)) return double.PositiveInfinity;
        double[] values = history.Where(item => now - item.At < HistoryMilliseconds).Select(item => item.Value).ToArray();
        if (values.Length < 3 || candidate && values.TakeLast(3).Any(value => !double.IsFinite(value)))
            return double.PositiveInfinity;
        double[] ok = values.Where(double.IsFinite).Order().ToArray();
        if (candidate && ok.Length < 3) return double.PositiveInfinity;
        if (ok.Length == 0) return 2000;
        double median = (ok[(ok.Length - 1) / 2] + ok[ok.Length / 2]) / 2;
        // P90 of eight samples is just the maximum: one old spike must not
        // count as several independent wins in overlapping sample windows.
        double p80 = ok[(int)Math.Ceiling(ok.Length * .8) - 1];
        return median + .75 * (p80 - median) + 500d * Math.Max(0, values.Length - ok.Length - 1) / values.Length;
    }

    private void Select(string path, long now) { _preferred = path; _selectedAt = now; ResetChallenge(); }
    private void ResetChallenge() { _challenger = null; _wins = 0; }
}
